### ISSUE TITLE
Follow up on behavior difference in `NullableAnalysis_24_State_From_NotType_Test_Class` and `NullableAnalysis_27_State_From_NotValue_Test_Class` unit-tests in `UnionsTests.cs`

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#83568]
Note a difference in diagnostics reported for Unions scenario and equivalent non-Unions scenario. Behavior for Unions scenario looks expected. The question is why do we get different behavior for non-Union scenario?

**Comment by AlekseyTs** [comment]
Here is a small scenario that I think is demonstrating an unexpected behavior:
```
#nullable enable

class S1
{
    public object? Value => throw null!;
}
class Program
{
    static void Test2(S1 s)
    {
        _ = s.Value;

        if (s.Value is int)
        {
#line 2000
            _ = s switch { { Value: {} } => 1 };
        }
    } 
}
```

Observed:
```
// (2000,19): warning CS8655: The switch expression does not handle some null inputs (it is not exhaustive). For example, the pattern 'null' is not covered.
//             _ = s switch { { Value: {} } => 1 };
Diagnostic(ErrorCode.WRN_SwitchExpressionNotExhaustiveForNull, "switch").WithArguments("null").WithLocation(2000, 19)
```

Expected: No warning reported.

**Comment by AlekseyTs** [comment]
After stepping through the compiler's code under debugger, it appears that the behavior is intentional. The `{}` pattern is treated as an explicit null test, which resets state of the property to "maybe null" prior to analyzing the switch exhaustiveness. I will update the affected tests and then close this issue.
